# Exploring Ebay Car Sales Data

In [10]:
import pandas as pd
import numpy as np
autos= pd.read_csv('autos.csv',encoding='Latin-1')

In [11]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

In [12]:
#changing the column names

autos.columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 
                 'ab_test','vehicle_type', 'registration_year', 'gearbox', 
                 'power_ps', 'model','odometer', 'registration_month', 
                 'fuel_type', 'brand','unrepaired_damage', 'ad_created', 
                 'num_photos', 'postal_code','last_seen']

In [13]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
date_crawled          50000 non-null object
name                  50000 non-null object
seller                50000 non-null object
offer_type            50000 non-null object
price                 50000 non-null object
ab_test               50000 non-null object
vehicle_type          44905 non-null object
registration_year     50000 non-null int64
gearbox               47320 non-null object
power_ps              50000 non-null int64
model                 47242 non-null object
odometer              50000 non-null object
registration_month    50000 non-null int64
fuel_type             45518 non-null object
brand                 50000 non-null object
unrepaired_damage     40171 non-null object
ad_created            50000 non-null object
num_photos            50000 non-null int64
postal_code           50000 non-null int64
last_seen             50000 non-null object
dtypes: int64(5)

In [14]:
autos.columns

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'ab_test',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'num_photos', 'postal_code',
       'last_seen'],
      dtype='object')

In [15]:
autos.describe(include='all')

date_crawled         name  seller offer_type  price ab_test  \
count                 50000        50000   50000      50000  50000   50000   
unique                48213        38754       2          2   2357       2   
top     2016-03-10 15:36:24  Ford_Fiesta  privat    Angebot     $0    test   
freq                      3           78   49999      49999   1421   25756   
mean                    NaN          NaN     NaN        NaN    NaN     NaN   
std                     NaN          NaN     NaN        NaN    NaN     NaN   
min                     NaN          NaN     NaN        NaN    NaN     NaN   
25%                     NaN          NaN     NaN        NaN    NaN     NaN   
50%                     NaN          NaN     NaN        NaN    NaN     NaN   
75%                     NaN          NaN     NaN        NaN    NaN     NaN   
max                     NaN          NaN     NaN        NaN    NaN     NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  num_photos   postal_code            last_seen  
count                 50000     50000.0  50000.000000                50000  
unique                   76         NaN           NaN                39481  
top     2016-04-03 00:00:00         NaN           NaN  2016-04-07 06:17:27  
freq                   1946         NaN           NaN                    8  
mean                    NaN         0.0  50813.627300                  NaN  
std                     NaN         0.0  25779.747957                  NaN  
min                     NaN         0.0   1067.000000                  NaN  
25%                     NaN         0.0  30451.000000                  NaN  
50%                     NaN         0.0  49577.000000                  NaN  
75%                     NaN         0.0  71540.000000                  NaN  
max                     NaN         0.0  99998.000000                  NaN

Findings:

1.Columns with mostly one values are: seller, offer_type, num_photos. These columns are candidates to be dropped, because no useful information can be extracted.
2.Numeric data stored as text: price, odometer.
3.Column ad_created needs further investigation, so far it seems that it consists of just one value
Let's drop three columns: seller, offer_type, num_photos

In [16]:
autos = autos.drop(columns=['seller','offer_type','num_photos'])

In [18]:
#converting column 'price' from object to numeric type
autos['price'] = (autos['price']
                 .str.replace('$','')
                 .str.replace(',','')
                  .astype(float)
                 )
#converting column 'odometer' from object to numeric type
autos['odometer'] = (autos['odometer']
                    .str.replace(',','')
                    .str.replace('km','')
                    .astype(float)
                    )

In [19]:
#renaming the column 'odometer' to 'odometer_km'
autos = autos.rename({'odometer':'odometer_km'}, axis=1)

In [20]:
print(autos['price'].unique().shape)
print(autos['price'].describe())
print(autos['price'].value_counts().head(20))

(2357,)
count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64
0.0       1421
500.0      781
1500.0     734
2500.0     643
1200.0     639
1000.0     639
600.0      531
800.0      498
3500.0     498
2000.0     460
999.0      434
750.0      433
900.0      420
650.0      419
850.0      410
700.0      395
4500.0     394
300.0      384
2200.0     382
950.0      379
Name: price, dtype: int64


Findings:

-There are 2357 unique values in column price.
-Around 3% of prices equal 0, these rows might be considered for removing from the dataset.
-The minimum price is 0 and maximum price is one hundred million dollars, they look too small and too large respectively.
Let's explore the lowest and the highest prices further on.

In [21]:
autos['price'].value_counts().sort_index(ascending=False).head(20)

99999999.0    1
27322222.0    1
12345678.0    3
11111111.0    2
10000000.0    1
3890000.0     1
1300000.0     1
1234566.0     1
999999.0      2
999990.0      1
350000.0      1
345000.0      1
299000.0      1
295000.0      1
265000.0      1
259000.0      1
250000.0      1
220000.0      1
198000.0      1
197000.0      1
Name: price, dtype: int64

In [22]:
autos['price'].value_counts().sort_index(ascending=True).head(20)

0.0     1421
1.0      156
2.0        3
3.0        1
5.0        2
8.0        1
9.0        1
10.0       7
11.0       2
12.0       3
13.0       2
14.0       1
15.0       2
17.0       3
18.0       1
20.0       4
25.0       5
29.0       1
30.0       7
35.0       1
Name: price, dtype: int64

Findings:

-A number of prices are below 30 dollars, the most frequent of them is 0 (1421 rows).
-Around 15 prices are higher than one million which seems too much.
-Starting from 350000 dollars, prices become twice higher.
-Considering the fact that prices on Ebay could start at 1$ because it is an auction site, 
I will keep prices from 1 dollar to 350000 dollars in the dataset and remove the rest.

In [23]:
autos = autos[autos['price'].between(1,350000)]
autos['price'].describe()

count     48565.000000
mean       5888.935591
std        9059.854754
min           1.000000
25%        1200.000000
50%        3000.000000
75%        7490.000000
max      350000.000000
Name: price, dtype: float64

In [24]:
autos['odometer_km'].value_counts().sort_index(ascending=False)

150000.0    31414
125000.0     5057
100000.0     2115
90000.0      1734
80000.0      1415
70000.0      1217
60000.0      1155
50000.0      1012
40000.0       815
30000.0       780
20000.0       762
10000.0       253
5000.0        836
Name: odometer_km, dtype: int64

Findings:

-More than half of vehicles have 150000 km on their odometer (65%). It is also the most frequent value (31414 rows)
-The mileage of the car and the number of published ads is in direct proportion
-There are no outliers

In [25]:
autos[['date_crawled','last_seen',
      'ad_created','registration_month',
      'registration_year']].info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 48565 entries, 0 to 49999
Data columns (total 5 columns):
date_crawled          48565 non-null object
last_seen             48565 non-null object
ad_created            48565 non-null object
registration_month    48565 non-null int64
registration_year     48565 non-null int64
dtypes: int64(2), object(3)
memory usage: 2.2+ MB


The columns registration_month and registration_year contain numeric values, while date_crawled, last_seen and ad_created represent strings.

First, I am going to explore string columns.

In [26]:
(autos['date_crawled']
 .str[:10]
 .value_counts(normalize=True,dropna=False)
 .sort_values()
)

2016-04-07    0.001400
2016-04-06    0.003171
2016-03-18    0.012911
2016-04-05    0.013096
2016-03-06    0.014043
2016-03-13    0.015670
2016-03-05    0.025327
2016-03-24    0.029342
2016-03-16    0.029610
2016-03-27    0.031092
2016-03-25    0.031607
2016-03-17    0.031628
2016-03-31    0.031834
2016-03-10    0.032184
2016-03-26    0.032204
2016-03-23    0.032225
2016-03-11    0.032575
2016-03-22    0.032987
2016-03-09    0.033090
2016-03-08    0.033296
2016-04-01    0.033687
2016-03-30    0.033687
2016-03-29    0.034099
2016-03-15    0.034284
2016-03-19    0.034778
2016-03-28    0.034860
2016-04-02    0.035478
2016-03-07    0.036014
2016-04-04    0.036487
2016-03-14    0.036549
2016-03-12    0.036920
2016-03-21    0.037373
2016-03-20    0.037887
2016-04-03    0.038608
Name: date_crawled, dtype: float64

Findings:

-The data was crawled every day within a month starting from 5 March 2016 and ending at 7 April 2016
-The distribution of listings crawled on each day is roughly uniform.

In [27]:
(autos['last_seen']
 .str[:10]
 .value_counts(normalize=True,dropna=False)
 .sort_index()
)

2016-03-05    0.001071
2016-03-06    0.004324
2016-03-07    0.005395
2016-03-08    0.007413
2016-03-09    0.009595
2016-03-10    0.010666
2016-03-11    0.012375
2016-03-12    0.023783
2016-03-13    0.008895
2016-03-14    0.012602
2016-03-15    0.015876
2016-03-16    0.016452
2016-03-17    0.028086
2016-03-18    0.007351
2016-03-19    0.015834
2016-03-20    0.020653
2016-03-21    0.020632
2016-03-22    0.021373
2016-03-23    0.018532
2016-03-24    0.019767
2016-03-25    0.019211
2016-03-26    0.016802
2016-03-27    0.015649
2016-03-28    0.020859
2016-03-29    0.022341
2016-03-30    0.024771
2016-03-31    0.023783
2016-04-01    0.022794
2016-04-02    0.024915
2016-04-03    0.025203
2016-04-04    0.024483
2016-04-05    0.124761
2016-04-06    0.221806
2016-04-07    0.131947
Name: last_seen, dtype: float64

Findings:

-The distribution above shows the percent of removed ads per each day. It looks like in the last three day the percent of removed ads was the highest, however it is not necessarily because of the increased number of cars sold.

-It is impossible to make any definite conclusions from this distribution.

In [28]:
print(autos["ad_created"].str[:10].unique().shape)
(autos["ad_created"]
        .str[:10]
        .value_counts(normalize=True, dropna=False)
        .sort_index()
        )

(76,)


2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
2015-12-30    0.000021
2016-01-03    0.000021
2016-01-07    0.000021
2016-01-10    0.000041
2016-01-13    0.000021
2016-01-14    0.000021
2016-01-16    0.000021
2016-01-22    0.000021
2016-01-27    0.000062
2016-01-29    0.000021
2016-02-01    0.000021
2016-02-02    0.000041
2016-02-05    0.000041
2016-02-07    0.000021
2016-02-08    0.000021
2016-02-09    0.000021
2016-02-11    0.000021
2016-02-12    0.000041
2016-02-14    0.000041
2016-02-16    0.000021
2016-02-17    0.000021
2016-02-18    0.000041
2016-02-19    0.000062
2016-02-20    0.000041
2016-02-21    0.000062
                ...   
2016-03-09    0.033151
2016-03-10    0.031895
2016-03-11    0.032904
2016-03-12    0.036755
2016-03-13    0.017008
2016-03-14    0.035190
2016-03-15    0.034016
2016-03-16    0.030125
2016-03-17    0.031278
2016-03-18    0.013590
2016-03-19    0.033687
2016-03-20    0.037949
2016-03-21 

Findings:

This distribution shows the percent of created ads per each date. The variety is large, it seems that the majority of ads were created in March, however some ads are much older, up to 10 months

In [29]:
autos['registration_year'].describe()

count    48565.000000
mean      2004.755421
std         88.643887
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

Findings:

-The minimum and maximum values don't seem to be true, therefore it is impossible to consider the calculations above for further analysis.
-Further investigation of this column is required. Let's calculate the percentage of data with invalid registration year.

In [30]:
autos = autos[autos['registration_year'].between(1900,2018)]
autos['registration_year'].value_counts(normalize=True).head(20)

2000    0.065015
2005    0.060482
1999    0.059679
2004    0.055683
2003    0.055600
2006    0.055003
2001    0.054302
2002    0.051212
1998    0.048678
2007    0.046907
2008    0.045630
2009    0.042952
1997    0.040191
2011    0.033434
2010    0.032734
2017    0.028676
1996    0.028284
2012    0.026986
1995    0.025277
2016    0.025132
Name: registration_year, dtype: float64

In [31]:
autos['brand'].value_counts(normalize=True)

volkswagen        0.212822
bmw               0.108646
opel              0.108646
mercedes_benz     0.095791
audi              0.085862
ford              0.069670
renault           0.047896
peugeot           0.029458
fiat              0.025998
seat              0.018932
skoda             0.016068
nissan            0.015265
mazda             0.015224
smart             0.014297
citroen           0.014111
toyota            0.012587
hyundai           0.009950
sonstige_autos    0.009641
volvo             0.009044
mini              0.008611
mitsubishi        0.008178
honda             0.007972
kia               0.007107
alfa_romeo        0.006592
porsche           0.005912
suzuki            0.005892
chevrolet         0.005665
chrysler          0.003481
dacia             0.002657
daihatsu          0.002513
jeep              0.002204
subaru            0.002101
land_rover        0.002039
saab              0.001627
daewoo            0.001566
jaguar            0.001524
trabant           0.001380
r

German brands represent four out of five top brands, in total it is almost 50% of our dataset. The most popular brand is Volkswagen, the next two brands are twice less presented.

Some brands don't have a significant percentage so they may be excluded from dataset and only those brands with percentage of 5 percent or higher will be left for further analysis.

In [32]:
brand_counts = autos['brand'].value_counts(normalize=True)
common_brands = brand_counts[brand_counts > .05].index
print(common_brands)

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford'], dtype='object')


In [33]:
brand_mean_prices = {}

for b in common_brands:
    b_only = autos[autos['brand'] == b]
    mean_price = b_only['price'].mean()
    brand_mean_prices[b] = int(mean_price)

brand_mean_prices

{'audi': 9212,
 'bmw': 8261,
 'ford': 3728,
 'mercedes_benz': 8526,
 'opel': 2941,
 'volkswagen': 5333}

Findings:

-The most expensive brand is audi
-The least expensive is opel
-volkswagen has the average price comparing to others, and if considering the fact that it has the greatest value of posted ads

In [35]:
brand_mean_mileage = {}
for b in common_brands:
    b_only = autos[autos['brand'] == b]
    mean_mil = b_only['odometer_km'].mean()
    brand_mean_mileage[b] = int(mean_mil)
    

mean_mileage = pd.Series(brand_mean_mileage).sort_values(ascending=False)
mean_price = pd.Series(brand_mean_prices).sort_values(ascending=False)
brand_agg = pd.DataFrame(mean_mileage,columns=['mean_mileage'])
brand_agg

mean_mileage
bmw                  132682
mercedes_benz        130848
audi                 129492
opel                 129452
volkswagen           128955
ford                 124349

In [36]:
brand_agg['mean_price'] = mean_price
brand_agg

mean_mileage  mean_price
bmw                  132682        8261
mercedes_benz        130848        8526
audi                 129492        9212
opel                 129452        2941
volkswagen           128955        5333
ford                 124349        3728

Findings:

-The more expensive the car, the greater value of mileage it has (with the exception of opel)
Next steps:

-Identify categorical data that uses german words in the dataset, translate them and map the values to their english counterparts

In [37]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

    price  ab_test vehicle_type  registration_year    gearbox  power_ps  \
0  5000.0  control          bus               2004    manuell       158   
1  8500.0  control    limousine               1997  automatik       286   
2  8990.0     test    limousine               2009    manuell       102   
3  4350.0  control   kleinwagen               2007  automatik        71   
4  1350.0     test        kombi               2003    manuell         0   

    model  odometer_km  registration_month fuel_type       brand  \
0  andere     150000.0                   3       lpg     peugeot   
1     7er     150000.0                   6    benzin         bmw   
2    golf      70000.0                   7    benzin  volkswagen   
3  fortwo      70000.0                   6    benzin       smart   
4   focus     150000.0                   7    benzin        ford   

  unrepaired_damage           ad_created  postal_code            last_seen  
0              nein  2016-03-26 00:00:00        79588  2016-04-06 06:45:54  
1              nein  2016-04-04 00:00:00        71034  2016-04-06 14:45:08  
2              nein  2016-03-26 00:00:00        35394  2016-04-06 20:15:37  
3              nein  2016-03-12 00:00:00        33729  2016-03-15 03:16:28  
4              nein  2016-04-01 00:00:00        39218  2016-04-01 14:38:50

it looks like data in German is presented in the following columns:

vehicle_type
gearbox
fuel_type
unrepaired_damage
Let's explore each column.

In [38]:
autos['vehicle_type'].unique()

array(['bus', 'limousine', 'kleinwagen', 'kombi', nan, 'coupe', 'suv',
       'cabrio', 'andere'], dtype=object)

In [39]:
autos['gearbox'].unique()

array(['manuell', 'automatik', nan], dtype=object)

In [40]:
autos['fuel_type'].unique()

array(['lpg', 'benzin', 'diesel', nan, 'cng', 'hybrid', 'elektro',
       'andere'], dtype=object)

In [41]:
autos['unrepaired_damage'].unique()

array(['nein', nan, 'ja'], dtype=object)

In order to translate the German words in the columns mentioned above, I will create a dictionary, with German words as keys and their English translations as values

In [42]:
words_translated = {
    'bus':'bus',
    'limousine':'limousine',
    'kleinwagen':'supermini',
    'kombi':'station_wagon',
    'coupe':'coupe',
    'suv':'suv',
    'cabrio':'cabrio',
    'andere' :'other',
    'manuell':'manual',
    'automatik':'automatic',
    'lpg':'lpg',
    'benzin':'petrol',
    'diesel':'diesel',
    'cng':'cng',
    'hybrid':'hybrid',
    'elektro':'electro',
    'nein':'no',
    'ja':'yes'
}
for each in ['vehicle_type','gearbox','fuel_type','unrepaired_damage']:
    autos[each] = autos[each].map(words_translated)

In [43]:
print(autos['vehicle_type'].unique())
print(autos['gearbox'].unique())
print(autos['fuel_type'].unique())
print(autos['unrepaired_damage'].unique())

['bus' 'limousine' 'supermini' 'station_wagon' nan 'coupe' 'suv' 'cabrio'
 'other']
['manual' 'automatic' nan]
['lpg' 'petrol' 'diesel' nan 'cng' 'hybrid' 'electro' 'other']
['no' nan 'yes']


task completed